<?xml version="1.0" encoding="ISO-8859-15"?>
<launch>

  <arg name="output_folder"           default="/media/adaptation/New_volume/Domain_Adaptation_Pipeline/IO_pipeline/Pipeline/Output_kimera_mesh" />
  <arg name="idtf"                    default="output" />
  <arg name="fps"                    default="10"/>
  <arg name="frame_limit"            default="-1"/>
  <arg name="ratio_reprojected"      default="1.0"/>
  <arg name="mapping_scannet_path"   default="/media/adaptation/New_volume/Domain_Adaptation_Pipeline/IO_pipeline/Scannet_DB/scannetv2-labels.combined.tsv"/>
  <arg name="scannet_labels"         default="False" />
  <arg name="voxel_size"             default="0.03"/>
  <arg name="max_ray_length_m"       default="5"/>
  <arg name="verbose" default="false"/>

  <!--  FRAMES  -->
  <arg name="sensor_frame"       default="base_link_forward"/>
  <arg name="base_link_gt_frame" default="base_link_forward"/>
  <arg name="base_link_frame"    default="base_link_forward"/>

  <!--  TOPICS  -->
  <arg name="depth_topic"       default="depth_img"/>
  <arg name="image_topic"       default="rgb_img"/>
  <arg name="seg_topic"         default="semantic_img"/>
  <arg name="sync_topic"        default="sync_semantic"/>
  <arg name="aux_labels"        default="invalid"/>
  <arg name="prob_aux"        default="0"/>
  <arg name="prob_main"        default="0"/>
  <arg name="semantic_label_2_color_csv_filename" default="nyu40_segmentation_mapping.csv"/>
  <arg name="use_dynamic_masking"         default="true"/>
  <arg name="semantic_pointcloud"         default="/semantic_pointcloud"/>
  <arg name="use_freespace_pointcloud"    default="false" />
  <arg name="freespace_pointcloud"        default="/dev/null"/>
  
  <node pkg="tf2_ros"
      type="static_transform_publisher"
      name="base_link"
      args="0 0 0 -0.5 0.5 -0.5 0.5 base_link base_link_forward" />

  <!-- 
  <node pkg="tf2_ros"
      type="static_transform_publisher"
      name="world_map"
      args="0 0 0 0 0 0 1 world map" /> 
            x y z qx qy qz qw frame_id child_frame_id
  -->

  <node name="kimera_interfacer" pkg="kimera_interfacer"
      type="kimera_interfacer" output="screen" required="true" >

      <!--Generate Semantic Pointcloud-->
      <remap from="pointcloud"                to="$(arg semantic_pointcloud)"/>

      <!--Kimera Tsdf Node Parameters-->
      <param name="base_link_gt_frame"        value="$(arg base_link_gt_frame)"/>
      <param name="base_link_frame"           value="$(arg base_link_frame)"/>
      <param name="world_frame"               value="map"/>
      <param name="verbose" value="$(arg verbose)"/>

      <!--Input-->
      <remap from="pointcloud"                to="$(arg semantic_pointcloud)"/>
      <param name="sync_topic" value="$(arg sync_topic)"/>

      <!--Params-->
      <param name="tsdf_voxel_size"           value="$(arg voxel_size)" />
      <param name="tsdf_voxels_per_side"      value="32" />
      <param name="max_ray_length_m"          value="$(arg max_ray_length_m)" />
      <param name="min_time_between_msgs_sec" value="0.2" />
      <param name="voxel_carving_enabled"     value="true" />
      <param name="color_mode"                value="lambert_color"/>
      <param name="use_const_weight"          value="false" />
      <param name="use_freespace_pointcloud"  value="$(arg use_freespace_pointcloud)" />
      <remap from="freespace_pointcloud"      to="$(arg freespace_pointcloud)"/>

      <param name="sensor_frame"              value="$(arg sensor_frame)"/>
      <param name="use_tf_transforms"         value="true" />

      <param name="enable_icp"                value="false" />
      <param name="icp_iterations"            value="10" />
      
      <!--Method to update voxels' information: "fast" or "merged"-->
      <param name="method"                     value="fast" />

      <!--"color", "semantic" or "semantic_probability"-->
      <param name="semantic_color_mode"              value="semantic"/>
      <param name="semantic_measurement_probability" value="0.8" />

      <!--The only dynamic label we have right now are humans, with label 20-->
      <rosparam param="dynamic_semantic_labels">[255]</rosparam>

      <param name="slice_level" value="1.0" />
      <param name="semantic_label_2_color_csv_filepath"  value="$(find kimera_interfacer)/cfg/$(arg semantic_label_2_color_csv_filename)"/>

      <param name="mesh_filename"           value="$(arg output_folder)/$(arg idtf)_predict_mesh.ply" />
      <param name="tsdf_filename"           value="$(arg output_folder)/$(arg idtf)_predict_tsdf_esdf_layers.vxblx" />
      <param name="semantic_filename"       value="$(arg output_folder)/$(arg idtf)_serialized.data" />
  </node>
</launch>
